M2/Macaulay2/m2/lists.m2

```diff
@@ -247,19 +247,17 @@ randomSubset(ZZ, ZZ) := (N, n) -> (
 	    while (N - t) * rawRandomRRUniform defaultPrecision >= n - m
 	    do t += 1;
 	    first (t, t += 1))))
--- when size not given, use binomial distribution first to determine size
 randomSubset ZZ := N -> (
     if N < 0 then error "expected a nonnegative integer"
     else if N < 30 then (
-	-- for small N, sum up N Bernoulli variates
-	n := 0;
-	scan(N, i -> n += random 2))
+	for i from 0 to N - 1 list if random 2 == 1 then i else continue)
```

Comment:
I think this algorithm is faster even for N>30. Did you check?

Reply:
Yeah, I did.  It's about the same.

Reply:
but this one is so much simpler and natural...

Reply:
I think simpler (perhaps even faster) is to generate a (64-bit) integer and use the bits to return integers. For N < 64 this should be near instantaneous. For larger N, split it into 64-bit blocks ...

Reply:
Ooh, I like the bit idea.  And with GMP, we're not even limited to 64 bits...

Reply:
Okay yeah, that's way faster.  Thanks!

Reply:
yeah `random 2` is slower than `random RR`. Using bits of a random integer is the way to go.

Reply:
Here's the new timing comparison (using the random bits is the last column).  Much nicer :)

```m2
      +----+-------+--------+--------+
o20 = |30  |.10018 |.0846158|.0232665|
      +----+-------+--------+--------+
      |50  |.170638|.136678 |.0367224|
      +----+-------+--------+--------+
      |100 |.341813|.254995 |.0783433|
      +----+-------+--------+--------+
      |200 |.695711|.624376 |.144193 |
      +----+-------+--------+--------+
      |500 |1.7485 |1.42215 |.410996 |
      +----+-------+--------+--------+
      |1000|3.41022|2.79286 |.858254 |
      +----+-------+--------+--------+
```